tweet_recommendations/our_method.py

```diff
@@ -0,0 +1,286 @@
+                                                           for item in t])
+
+        if self.embedding_model is not None:
+            x[self.embedding_name] = self.embed_lemmas(x['lemmas'])
+        else:
+            assert len(x[self.embedding_name].iloc[0].shape) == 1, \
+                f"`{self.embedding_name}` column should contain single vectors."
+
+        if self.verbose:
+            print("Dropping tweets without proper words.")
+            count_before = len(x)
+        x = x.dropna(subset=[self.embedding_name])
+        if self.verbose:
+            print(f"Dropped: {count_before - len(x)} rows.")
+        if self.verbose:
+            print('Building Graph')
+        g = self._build_base_graph(x)
+
+        if self.verbose:
+            print('Calculating hashtag embeddings.')
+        g = self._calculate_hashtag_embeddings(g)
+
+        if self.verbose:
+            print('Calculating popularity measures.')
+        g = self._calculate_hashtag_mean_retweets(g)
+        g = self._calculate_edge_weights(g)
+        self._G = self._calculate_pagerank(g)
+
+        hashtags = [{'hashtag': node, **self._G.nodes[node]}
+                    for node in self._G.nodes
+                    if self._G.nodes[node]['node_type'] == 'hashtag']
+        self._hashtags_df = pd.DataFrame(hashtags)
+
+    def embed_lemmas(self, texts_lemmas):
+        if not isinstance(texts_lemmas, pd.Series):
+            texts_lemmas = pd.Series(texts_lemmas)
+
+        def _embed_text(lemmas):
+            vectors = []
+            for word in lemmas:
+                try:
+                    vectors.append(self.embedding_model.word_vec(word.lower()))
+                except KeyError:
+                    continue
+            if len(vectors) > 0:
+                return np.asarray(vectors).mean(axis=0)
+            else:
+                return None
+        if self.verbose:
+            print("Embedding lemmas")
+            tqdm.tqdm.pandas()
+            return texts_lemmas.progress_apply(_embed_text)
+        else:
+            return texts_lemmas.apply(_embed_text)
+
+    def transform(self, x: Union[Tuple[Tuple[str, ...]], Tuple[str, ...],
+                                 Tuple[np.ndarray]],
+                  **transform_params):
+
+        x = list(x)
+        # tokenize and lemmatize text
+        if isinstance(x[0], str):
+            x = [get_wcrft2_results_for_text(txt) for txt in x]
+
+        # turn tokens to embeddings
+        if isinstance(x[0], list):
+            x = list(self.embed_lemmas(x))
+
+        # calculate mean tweet embeddings
+        if len(x[0].shape) > 1:
+            x = [emb.mean(axis=0) for emb in x]
+
+        hashtag_emb = np.asarray(self._hashtags_df[self.embedding_name].tolist())
+
+        if self.verbose:
+            print("Calculating similarities")
+        sim = self.embedding_similarity(np.asarray(x), hashtag_emb)
+        sim = ((sim - sim.min(axis=1).reshape(-1, 1))

```

Comment:
Napisałeś metodę `normalise`, w której robisz dokładnie to samo, a jej nie używasz. Możesz ją zgeneralizować, żeby dodatkowo przyjmowała parametr `axis`.

Reply:
Nie do końca robią to samo, bo `normalise` jeśli min i max są "równe" nie zmienia wejścia i zwraca je takie jakie było. 
A tutaj jak min i max są równe to zwróci wektor samych 0. Muszę pomyśleć czy wprowadzenie w obu miejscach takiego samego zachowania nie spowoduje zmiany działania metody.

Reply:
Ujednolicone. Jeśli np wszystkie hashtagi są równoodległe od zapytania, to nie ma dla ustalania rankingu znaczenia czy znormalizowana odległość wyniesie 0  czy 1 czy 0.5 czy cokolwiek innego.